staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/implementation.go

```diff
@@ -0,0 +1,314 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/interpreter"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	"k8s.io/api/admissionregistration/v1alpha1"
+	authenticationv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	api_types "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/plugin/cel/matching"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/kube-openapi/pkg/common"
+)
+
+var _ ValidatorCompiler = &CValidatorCompiler{}
+var _ matching.MatchCriteria = &matchCriteria{}
+
+type matchCriteria struct {
+	constraints *v1alpha1.MatchResources
+}
+
+func (m *matchCriteria) GetParsedNamespaceSelector() (labels.Selector, error) {
+	selector, err := metav1.LabelSelectorAsSelector(m.constraints.NamespaceSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	return selector, nil
+}
+
+func (m *matchCriteria) GetParsedObjectSelector() (labels.Selector, error) {
+	selector, err := metav1.LabelSelectorAsSelector(m.constraints.ObjectSelector)
+	if err != nil {
+		return nil, err
+	}
+
+	return selector, nil
+}
+
+func (m *matchCriteria) GetMatchResources() v1alpha1.MatchResources {
+	return *m.constraints
+}
+
+type CValidatorCompiler struct {
+	Matcher *matching.Matcher
+}
+
+func (c *CValidatorCompiler) SetExternalKubeInformerFactory(factory informers.SharedInformerFactory) {
+	c.Matcher.SetNamespaceLister(factory.Core().V1().Namespaces().Lister())
+}
+
+func (c *CValidatorCompiler) SetExternalKubeClientSet(k kubernetes.Interface) {
+	c.Matcher.SetExternalKubeClientSet(k)
+}
+
+func (c *CValidatorCompiler) DefinitionMatches(definition *v1alpha1.ValidatingAdmissionPolicy, a admission.Attributes, o admission.ObjectInterfaces) (bool, error) {
+	criteria := matchCriteria{constraints: definition.Spec.MatchConstraints}
+	return c.Matcher.Matches(&criteria, a, o)
+}
+
+func (c *CValidatorCompiler) BindingMatches(binding *v1alpha1.ValidatingAdmissionPolicyBinding, a admission.Attributes, o admission.ObjectInterfaces) (bool, error) {
+	if binding.Spec.MatchResources == nil {
+		return true, nil
+	}
+	criteria := matchCriteria{constraints: binding.Spec.MatchResources}
+	return c.Matcher.Matches(&criteria, a, o)
+}
+
+func (c *CValidatorCompiler) ValidateInitialization() error {
+	if err := c.Matcher.ValidateInitialization(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+type validationActivation struct {
+	object    *unstructured.Unstructured
+	oldObject *unstructured.Unstructured
+	params    *unstructured.Unstructured
+	request   *unstructured.Unstructured
+	hasOld    bool
```

Comment:
I don't think hasOld is needed (see above comment)

Reply:
Did this get resolved on accident? `hasOld` is still present.

Reply:
It has been removed. Sorry